lib/improver/utilities/statistical_operations.py

```diff
@@ -0,0 +1,213 @@
+        """
+        self.reference_cube = reference_cube
+        self.output_name = output_name
+        self.inverse_ordering = inverse_ordering
+
+    def create_probability_cube(self, cube):
+        """
+        Create a probability cube in which to store the calculated
+        probabilities.
+
+        Args:
+            cube (iris.cube.Cube):
+                The input cube to be used as a template.
+        Returns:
+            probability_cube (iris.cube.Cube):
+                The new cube with suitable metadata.
+        """
+        cube_format = next(cube.slices([cube.coord(axis='y'),
+                                        cube.coord(axis='x')]))
+        probabilities = cube_format.copy(data=np.full(cube_format.shape, 1.,
+                                                      dtype=float))
+        percentile_coordinate, = [coord.name() for coord in
+                                  cube_format.coords()
+                                  if 'percentile' in coord.name()]
+        if percentile_coordinate:
+            probabilities.remove_coord(percentile_coordinate)
+        probabilities.units = 1
+        probabilities.rename(self.output_name)
+        return probabilities
+
+    def percentile_interpolation(self, cube, reference_cube):
+        """
+        Perform the percentile interpolation to construct the probability
+        field.
+
+        Args:
+            cube (iris.cube.Cube):
+                A cube of values, that effectively behave as thresholds, for
+                which it is desired to obtain probability values from a
+                percentiled reference cube.
+        Returns:
+            probabilities (iris.cube.Cube):
+                A cube of probabilities obtained by interpolating the
+                percentile values.
+        """
+        probabilities = self.create_probability_cube(reference_cube)
+
+        array_shape = list(cube.shape)
+        array_shape.insert(0, 2)
+        array_shape = tuple(array_shape)
+        percentile_bounds = np.full(array_shape, -1, dtype=float)
+        height_bounds = np.full(array_shape, -1001., dtype=float)
+        height_bounds[1] = -1.
+
+        percentile_coordinate, = [coord for coord in reference_cube.coords()
+                                  if 'percentiles' in coord.name()]
+        percentiles = percentile_coordinate.points
+
+        for index, pslice in enumerate(reference_cube.slices_over(
+                percentile_coordinate)):
+            indices = (cube.data < pslice.data if self.inverse_ordering else
+                       cube.data > pslice.data)
+            percentile_bounds[0, indices] = percentiles[index]
+            height_bounds[0, indices] = pslice.data[indices]
+            try:
+                # Usual behaviour where the orography falls between heights
+                # corresponding to percentiles.
+                percentile_bounds[1, indices] = percentiles[index + 1]
```

Comment:
What happens if several percentiles have the same value

Reply:
As a coordinate, I think the percentile value (ie on the 0-100 spectrum) should not be multiply defined?  If the height value is the same for several different percentile levels, this is caught by ignoring the divide-by-zero in calculating the interpolant.  The output probability then relates to the lowest of the possible percentiles - eg if a height of 4.0 km corresponded to the 0th and 10th percentiles, the code as written would assign that a probability of "0" (0.1 if across the 10th and 20th percentiles, etc).  But there are no problems with singularity or NANs.

I've added a unit test for this behaviour.  We may need to think about whether this is actually what we want, and what the ideal behaviour would look like.